### PR TITLE
7zz: compile optimized assembly code for x86_64

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

This PR compiles `7zz` with optimized assembly code for faster operations. To do this, this PR also introduces the `uasm` compiler at version `2.53`.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by thiagokokada** [comment]
Comparison.

Before:
```
7-Zip (z) 21.07 (x64) : Copyright (c) 1999-2021 Igor Pavlov : 2021-12-26
 64-bit locale=pt_BR.UTF-8 Threads:6

Scanning the drive for archives:
1 file, 207936691 bytes (199 MiB)

Extracting archive: [GDI] Capcom vs. SNK 2 - Millionaire Fighting 2001 (JP)(en).7z

WARNINGS:
There are data after the end of archive

--
Path = [GDI] Capcom vs. SNK 2 - Millionaire Fighting 2001 (JP)(en).7z
Type = 7z
WARNINGS:
There are data after the end of archive
Physical Size = 207936653
Tail Size = 38
Headers Size = 279
Method = LZMA:24
Solid = +
Blocks = 1

Everything is Ok

Archives with Warnings: 1

Warnings: 1
Files: 4
Size:       1188202263
Compressed: 207936691
7zz x   10.83s user 0.40s system 98% cpu 11.439 total
```

After:
```
7-Zip (z) 21.07 (x64) : Copyright (c) 1999-2021 Igor Pavlov : 2021-12-26
 64-bit locale=pt_BR.UTF-8 Threads:6, ASM

Scanning the drive for archives:
1 file, 207936691 bytes (199 MiB)

Extracting archive: [GDI] Capcom vs. SNK 2 - Millionaire Fighting 2001 (JP)(en).7z

WARNINGS:
There are data after the end of archive

--
Path = [GDI] Capcom vs. SNK 2 - Millionaire Fighting 2001 (JP)(en).7z
Type = 7z
WARNINGS:
There are data after the end of archive
Physical Size = 207936653
Tail Size = 38
Headers Size = 279
Method = LZMA:24
Solid = +
Blocks = 1

Everything is Ok

Archives with Warnings: 1

Warnings: 1
Files: 4
Size:       1188202263
Compressed: 207936691
/nix/store/g95spkz5s5vdx3yvzrjaq1hvpbhr8snp-7zz-21.07/bin/7zz x   6.87s user 0.40s system 98% cpu 7.395 total
```

**Comment by thiagokokada** [comment]
CC @peterhoeg.

**Comment by thiagokokada** [comment]
@ofborg build _7zz

File: pkgs/development/compilers/uasm/default.nix [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-Aohwrcb/KTKUFFpfmqVDPNjJh1dMYSNnBJ2eFaP20pM=";
+  };
+
+  patches = [
+    # fix `dbgcv.c:*:*: fatal error: direct.h: No such file or directory`
+    (fetchpatch {
+      name = "fix_build_dbgcv_c_on_unix.patch";
+      url = "https://patch-diff.githubusercontent.com/raw/Terraspace/UASM/pull/154.patch";
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/uasm/default.nix:18]: That isn't merged yet so we need to fetch each commit separate.

Comment (thiagokokada) [reply pkgs/development/compilers/uasm/default.nix]: Fixed.

**Comment by thiagokokada** [comment]
@ofborg build _7zz

**Comment by peterhoeg** [comment]
Minor thing. I suggest you introduce a `useUasm = stdenv.isx86_64;` variable and do the conditionals based on that instead. It communicates intent better as *technically* you set `MY_ASM=uasm` because you want to use `uasm` - not because you're on 64bit Linux .

**Comment by thiagokokada** [comment]
> Minor thing. I suggest you introduce a `useUasm = stdenv.isx86_64;` variable and do the conditionals based on that instead. It communicates intent better as _technically_ you set `MY_ASM=uasm` because you want to use `uasm` - not because you're on 64bit Linux .

Done. Had to refactor some parts of the code, otherwise the `platformSuffix` would look very strange with a `if useUasm then "_x64"`. At least, in case there are more platforms that have assembly optimization in the future, it should be easier to support now.

File: pkgs/tools/archivers/7zz/default.nix [hunk 2]
```diff
@@ -1,5 +1,14 @@
-{ stdenv, lib, fetchurl, p7zip }:
-
+{ stdenv, lib, fetchurl, p7zip, uasm, useUasm ? stdenv.isx86_64 }:
+
+let
+  inherit (stdenv.hostPlatform) system;
+  platformSuffix =
+    if useUasm then
```
Comment (SuperSandro2000) [inline pkgs/tools/archivers/7zz/default.nix:6]: We could also use lib.optionalString here but doesn't matter to much.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 157305` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
2 packages built:
</summary>
<ul>
<li>_7zz</li>
<li>uasm</li>
</ul>
</details>

**Pull request merged**